### DISCUSSION TITLE
Decorators on handler do not work with TestClient

### DISCUSSION THREAD

**Discussion by edmondop** [fastapi/fastapi#11884] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

Wrapping handlers works correctly and I can test my handler using curl
```python
   def _map_error(self, handler_func):
        ## See https://github.com/tiangolo/fastapi/issues/1743
        @functools.wraps(handler_func)
        async def wrapper(*args, **kwargs):
            try:
                return await handler_func(*args, **kwargs)
            except MyFooException as e:
                raise HTTPException(status_code=e.response.status_code, detail=str(e)) from e

        return wrapper

```

However, the TestClient invocation returns a 422  "Unprocessable Entity". This is the same error that is returned if you forget to use `functools.wrap` to preserve the signature of the handler when wrapping it

### Description

If you deploy a simple handler and you wrap it 

### Operating System

Linux, macOS

### FastAPI Version

0.111.1

### Pydantic Version

2.8.2

### Python Version

3.10.9

**Comment by YuriiMotov** [comment, marked as answer]
Could you provide a [MRE](https://en.wikipedia.org/wiki/Minimal_reproducible_example)?

In my attempt to reproduce the error I created the code below, but it works as intended:
```
import functools
from fastapi import FastAPI, HTTPException
from fastapi.testclient import TestClient

class ExcResponse():
    def __init__(self, status_code: int):
        self.status_code = status_code

class MyFooException(Exception):
    def __init__(self, status_code: int):
        self.response = ExcResponse(status_code=status_code)

def _map_error(handler_func):
    @functools.wraps(handler_func)
    async def wrapper(*args, **kwargs):
        try:
            return await handler_func(*args, **kwargs)
        except MyFooException as e:
            raise HTTPException(status_code=e.response.status_code, detail=str(e)) from e

    return wrapper

app = FastAPI()


@app.get("/")
@_map_error
async def index():
    raise MyFooException(status_code=401)


def test_decorator():
    client = TestClient(app=app)
    res = client.get("/")
    assert res.status_code == 401
```

**Comment by edmondop** [comment]
I realized the problem comes from using a class too. Without the `Component` class below, the problem doesn't occur

```python
from fastapi import FastAPI, HTTPException, APIRouter
from fastapi.testclient import TestClient
import uvicorn

import functools

app = FastAPI()


class ServiceException(Exception):
    pass


class UnderlyingService:
    async def do_stuff(self, input: str) -> str:
        return input[::-1]


class Component:
    def __init__(self, service: UnderlyingService):
        self.service = service
        self.router = APIRouter()
        self.router.add_api_route(
            "/do_stuff", _map_error(UnderlyingService().do_stuff), methods=["GET"]
        )

    def attach_to_app(self, app: FastAPI):
        app.include_router(self.router)


def _map_error(handler_func):
    ## See https://github.com/tiangolo/fastapi/issues/1743
    @functools.wraps(handler_func)
    async def wrapper(*args, **kwargs):
        try:
            return await handler_func(*args, **kwargs)
        except ServiceException as e:
            raise HTTPException(status_code=500, detail=str(e)) from e

    return wrapper


component = Component(UnderlyingService())
component.attach_to_app(app)

# uncomment this to see it failing
# if __name__ == "__main__":
#     client = TestClient(app)
#     response = client.get("/do_stuff?input=foo")
#     assert response.status_code == 200, response.status_code

if __name__ == "__main__":
    config = uvicorn.Config(app, host="0.0.0.0", port=8082, log_level="info")
    server = uvicorn.Server(config)
    server.run()
```
If you run a Vanilla curl, you see that it works

**Reply by YuriiMotov** [reply]
I tried your code with uncommented part. It doesn't fail. It works as expected.

In the initial question you said `Decorators on handler do not work with TestClient`. But how can you see it using the code above?

All that your decorator does is that it catches `ServiceException` and raises `HTTPException` in that case.
Since your code doesn't raise any `ServiceException`, you will not see any effect of `_map_error`.
